### PR TITLE
cow-rc-str: Use ptr::NonNull.

### PR DESCRIPTION
**Pull request by emilio**
This fixes UB caught by miri.

### CODE REVIEW

File: src/cow_rc_str.rs [hunk 1]
```diff
@@ -3,37 +3,24 @@
...
 use std::rc::Rc;
-use std::slice;
-use std::str;
+use std::{cmp, fmt, hash, marker, mem, ops, slice, str, ptr};
 
 /// A string that is either shared (heap-allocated and reference-counted) or borrowed.
 ///
 /// Equivalent to `enum { Borrowed(&'a str), Shared(Rc<String>) }`, but stored more compactly.
 ///
-/// FIXME(https://github.com/rust-lang/rfcs/issues/1230): use an actual enum if/when
```
[the remark this answers is not in the document]
Comment (glandium) [reply src/cow_rc_str.rs:20]: Mmmmm enum and Option<enum> don't appear to be the same size...

[the remark this answers is not in the document]
Comment (emilio) [reply src/cow_rc_str.rs:20]: Yeah, the enum would be larger unfortunately.

**Comment by glandium** [comment]
LGTM.

**Pull request merged**